### ISSUE TITLE
Unlimited items_per_page

### ISSUE THREAD

**Issue by felicitus** [api-platform/core#130]
Sometimes it's required to retrieve all items. Right now, I can't pass something like "itemsPerPage: -1" to indicate that I wish to retrieve all items.

This is probably a feature request. If that's already possible, it needs to be documented.

**Comment by dunglas** [comment]
I'm working on that new feature

**Comment by theofidry** [comment]
+1 ;)

**Comment by sroze** [comment]
Can we say that's the `0` value instead of `-1` that disable the pagination ?

**Comment by felicitus** [comment]
It would be best to configure the value to indicate that no pagination is wanted, i.e. -1, 0 or all, depending on the client library to use.

**Comment by sroze** [comment]
Configuration of a configuration value ?

**Comment by theofidry** [comment]
`-1`  is THE logical value for the "invalid" value in this case. `0` is still a valid value although there is no interest to get a collection without any member... So it feels more logical to me at least.

As @felicitus said, whatever the choice is for the default value, keeping it configurable would be a wise.

**Comment by sroze** [comment]
@theofidry It's FAR from being the logical value. A `-1` value (even the `0` value btw) for `itemsPerPage` makes no sense at all, except in some BE devs' head. I'd prefer a "pagination" strategy based on the pagination query parameter.

`/api/collection?pagination=false` would decorate the [getCollection](https://github.com/dunglas/DunglasApiBundle/blob/9f0bfc54bfff5a3d824951bbce5d8afb70a7bf2a/Doctrine/Orm/DataProvider.php#L95) method with an simple "WithoutPaginator" while `/api/collection?pagination[itemsPerPage]=10` would decorate it as [it's already the case](https://github.com/dunglas/DunglasApiBundle/blob/9f0bfc54bfff5a3d824951bbce5d8afb70a7bf2a/Doctrine/Orm/DataProvider.php#L125)

**Comment by theofidry** [comment]
If we keep with the `itemsPerPage` parameter, I find `-1` more logical than `0`. If we use an option `pagination` instead, I agree entirely with you, the `pagination=false` would make more sense, or `pagination=0` since it would be a boolean.

**Comment by dunglas** [comment]
The PR I started to work on is a bit different:

it add the following methods to `ResourceInterface`:
```php
/** @return bool */
public function isPaginationEnabled();

/** @return float */
public function getItemsPerPage();
```

Both can be configured globally (with a `CompilerPass`) and on a per resource basis in the service definition.

**Comment by sroze** [comment]
@dunglas So you can't disable pagination on-the-fly ?

**Comment by felicitus** [comment]
@dunglas can an API use enable/disable pagination on demand?

**Comment by dunglas** [comment]
@sroze not in my current implementation.

**Comment by dunglas** [comment]
@felicitus Why not. I don't get the use case for now but it should be ok to do that.
Another parameter such as `enable_pagination=0` would be better than a special value to `items_per_page`.

**Comment by dunglas** [comment]
See #130 for my proposal.

**Comment by felicitus** [comment]
I will test this tomorrow or on monday, but I guess I can modify ExtJS to respect enable_pagination.

**Comment by felicitus** [comment]
I just tried the PR and I can disable pagination in my config.yml, so that works for now. Have you already implemented the enable_pagination parameter?

**Comment by dunglas** [comment]
Not yet.

**Comment by pierrick-vignand** [comment]
In my opinion, collections should always be paginated to avoid huge request.
There is 206 HTTP status code for partial content.
Content-Range and Accept-Range headers define pagination properties.
Link headers indicate next page and previous page.
So, if your client application gets a 206 status code and you want to get all items, just call the next page.

A french article talk about that : http://blog.octo.com/designer-une-api-rest/#pagination

**Comment by dunglas** [comment]
The pagination is always enabled by default. No worries about that. However I'm working in #131 on a way to disable it via opt-in on a per `Resource` basis.

We have make the choice to use `PagedCollection` of the Hydra specification instead of HTTP headers: http://www.w3.org/ns/hydra/spec/latest/core/#collections
It's already fully implemented in DunglasApiBundle and more standard.

**Comment by pierrick-vignand** [comment]
ok, it looks great.

**Comment by theofidry** [comment]
Solved via #131 but the doc is still missing @dunglas?

**Comment by theofidry** [comment]
@sroze has the doc been updated?
